source/opt/type_manager.cpp

```diff
@@ -654,28 +646,32 @@ Type* TypeManager::RecordIfTypeDefinition(const Instruction& inst) {
       const Instruction* length_constant_inst = id_to_constant_inst_[length_id];
       assert(length_constant_inst);
 
-      // If it is a specialised constants, retrieve its SpecId.
+      // If it is a specialised constant, retrieve its SpecId.
+      // Only OpSpecConstantOp has a SpecId.
       uint32_t spec_id = 0u;
-      Type* length_type = nullptr;
-      Operand::OperandData length_words;
-      if (spvOpcodeIsSpecConstant(length_constant_inst->opcode())) {
+      if (length_constant_inst->opcode() == SpvOpSpecConstant) {
         context()->get_decoration_mgr()->ForEachDecoration(
             length_id, SpvDecorationSpecId,
             [&spec_id](const Instruction& decoration) {
               assert(decoration.opcode() == SpvOpDecorate);
               spec_id = decoration.GetSingleWordOperand(2u);
             });
-      } else {
-        length_type = GetType(length_constant_inst->type_id());
-        length_words = length_constant_inst->GetOperand(2u).words;
       }
+      // Determine extra words required to distinguish this array length
+      // from another.
+      std::vector<uint32_t> extra_words;
+      const auto opcode = length_constant_inst->opcode();
+      if (opcode == SpvOpConstant) {
+        extra_words.insert(extra_words.end(),
+                           length_constant_inst->GetOperand(2).words.begin(),
+                           length_constant_inst->GetOperand(2).words.end());
+      } else if (opcode == SpvOpSpecConstant) {
+        extra_words.push_back(spec_id);
```

Comment:
Can the same spec id be used in multiple decoration? I wonder about the case where the linker is attempting to link two modules with the same spec id, but different default values. What happens if such a module is not specialized?

Should we also include the default value in the extra words?

Reply:
Good point.  I've adde the literal value words in this case too.